odbc_fdw.c

```diff
@@ -1645,163 +1702,150 @@ odbcIterateForeignScan(ForeignScanState *node)
+					{
+						// no info about remaining data size; keep reading with same chunk_size
+						used_buffer_size += effective_chunk_size;
+					}
+					else
 					{
-						/* Unknown total size, must copy part by part */
-						accum_buffer_size = 0;
-						accum_buffer = NULL;
-						while (1)
+						// we read chunk_size, but there was result_size pending in total;
+						// adjust chunk_size for the remaining, so next wil hopely be the final chunk
+						used_buffer_size += effective_chunk_size;
+						// note that we need to read result_size - effective_chunk_size more data bytes,
+						chunk_size = (int)result_size - effective_chunk_size;
+						// wait, maybe we don't need to read, just append a zero!
+						if (chunk_size == 0)
 						{
-							size_t buf_len = buf[col_size] ? col_size + 1 : col_size;
-							// Allocate new accumulation buffer if necessary
-							if (accum_used + buf_len > accum_buffer_size)
+							if (!binary_data)
 							{
-								char *new_buff;
-								accum_buffer_size = accum_buffer_size == 0 ? col_size*2 : accum_buffer_size*2;
-								new_buff = (char *) palloc(sizeof(char) * (accum_buffer_size+1));
-								if (accum_buffer)
-								{
-									memmove(new_buff, accum_buffer, accum_used);
-									pfree(accum_buffer);
-								}
-								accum_buffer = new_buff;
-								accum_buffer[accum_used] = 0;
+								resize_buffer(&buffer, &buffer_size, used_buffer_size, used_buffer_size + 1);
+								buffer[used_buffer_size - 1] = 0;
 							}
-							// Copy part to the accumulation buffer
-							strncpy(accum_buffer+accum_used, buf, buf_len);
-							accum_used += buf_len;
-							accum_buffer[accum_used] = 0;
-							// Get new part
-							if (ret != SQL_SUCCESS_WITH_INFO)
-								break;
-							ret = SQLGetData(stmt, i, SQL_C_CHAR, buf, sizeof(char) * (col_size+1), &indicator);
-						};
-
+							break;
+						}
+						if (!binary_data)
+						{
+							chunk_size += 1;
+						}
 					}
-					else
+				}
+				else if (truncation == FRACTIONAL_TRUNCATION)
+				{
+					/* Fractional truncation has occurred;
+					* at this point we cannot obtain the lost digits
+					*/
+					used_buffer_size += effective_chunk_size;
+					if (chunk_size == effective_chunk_size)
 					{
-						/* We need to retrieve indicator more characters */
-						size_t buf_len = buf[col_size] ? col_size + 1 : col_size;
-						accum_buffer_size = buf_len + indicator;
-						accum_buffer = (char *) palloc(sizeof(char) * (accum_buffer_size+1));
-						strncpy(accum_buffer, buf, buf_len);
-						accum_buffer[buf_len] = 0;
-						ret = SQLGetData(stmt, i, SQL_C_CHAR, accum_buffer+buf_len, sizeof(char) * (indicator+1), &indicator);
+						/* The driver has omitted the trailing zero */
+						resize_buffer(&buffer, &buffer_size, used_buffer_size, used_buffer_size + 1);
+						buffer[used_buffer_size] = 0;
 					}
-					pfree(buf);
-					buf = accum_buffer;
+					elog(NOTICE,"Truncating number: %s",buffer);
```

Comment:
I don't think this message as is provides any information useful to the caller.

Reply:
Yep, I'll turn it into a debug message

Reply:
Actually, that's a nasty piece of code that I've kept from the previous version and which I don't know how to test properly